### PR TITLE
Support params in Factory in ITestResult

### LINKED ISSUE #1041: Factory dataprovider parameters not displayed in testresult
When I was running my tests using just a dataProvider on the methods, the used parameter was shown in the test result and this would nicely display in jenkins using the TestNG result plugin. I am now using a Factory with dataProvider to initialize my test class, but can no longer see the used parameter anywhere in test results. This is not so useful for my selenium tests, as I would like to know which instance failed (ex chrome could have non-standard behaviour and fail while IE does not). Not having this information available makes debugging test failures much harder.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #1041

Fixes #1041 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/ITestNGMethod.java [hunk 1]
```diff
@@ -211,4 +211,8 @@
   default boolean isDataDriven() {
     return false;
   }
+
+  default Object getRawInstance() {
```
Comment (juherr) [inline src/main/java/org/testng/ITestNGMethod.java]: Maybe some javadoc is missing. What's the difference between raw instance and instance?

File: src/main/java/org/testng/internal/BaseTestMethod.java [hunk 2]
```diff
@@ -715,4 +724,8 @@ public ConstructorOrMethod getConstructorOrMethod() {
   public String getQualifiedName() {
     return getRealClass().getName() + "." + getMethodName();
   }
+
+  public Object getRawInstance() {
```
Comment (juherr) [inline src/main/java/org/testng/internal/BaseTestMethod.java]: `@Override` is missing?

File: src/main/java/org/testng/internal/TestNGClassFinder.java [hunk 3]
```diff
@@ -34,7 +34,7 @@
 
   private static final String PREFIX = "[TestNGClassFinder]";
   private final ITestContext m_testContext;
-  private final Map<Class<?>, List<Object>> m_instanceMap = Maps.newHashMap();
+  private final Map<Class<?>, List<Object>>  m_instanceMap = Maps.newHashMap();
```
Comment (juherr) [inline src/main/java/org/testng/internal/TestNGClassFinder.java]: extra space added

**Review by juherr — COMMENTED** [review]
I don't catch why you use IParameterInfo.embeddedInstance everywhere instead of directly dealing with IParameterInfo.

**Comment by krmahadevan** [comment]
@juherr 

> I don't catch why you use IParameterInfo.embeddedInstance everywhere instead of directly dealing with IParameterInfo.

The reason behind doing that is because a user has the ability to retrieve the `instance` associated with a `ITestNGMethod` object. If we were to replace the underlying object with a `IParameterInfo` then we will affect backward compatibility.

The other option for this would be to deprecate the `getInstance()` method and introduce a new one that returns a `IParameterInfo` object. Should we go that route ? I chose not to, because there's not many people who are looking for this feature and I didn't want to introduce refactoring (Even if its a very minimal one wherein people will have to replace the call to `getInstance()` with a call to the new method that returns a `IParameterInfo` object).

File: src/main/java/org/testng/internal/TestNGClassFinder.java (hunk 3 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/TestNGClassFinder.java]: fixed

File: src/main/java/org/testng/internal/BaseTestMethod.java (hunk 2 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/BaseTestMethod.java]: fixed

File: src/main/java/org/testng/ITestNGMethod.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/ITestNGMethod.java]: Renamed this for better readability

File: src/main/java/org/testng/ITestNGMethod.java [hunk 4]
```diff
@@ -211,4 +212,12 @@
   default boolean isDataDriven() {
     return false;
   }
+
+  /**
+   * @return - A {@link IParameterInfo} object that represents details about the parameters
+   * associated with the factory method.
+   */
+  default IParameterInfo getFactoryMethodParamsInfo() {
```
Comment (juherr) [inline src/main/java/org/testng/ITestNGMethod.java]: Why not have the method on a IFactoryMethod which extends ITestNGMethod?

**Review by juherr — APPROVED** [review]

**Comment by krmahadevan** [comment]
Merging this PR since @juherr has approved it.

File: src/main/java/org/testng/ITestNGMethod.java (hunk 4 above)
Comment (juherr) [reply src/main/java/org/testng/ITestNGMethod.java]: Please, create an issue as reminder

**Pull request merged**